### DISCUSSION TITLE
Problem with converting CoNLL-U annotations into spaCy Docs

### DISCUSSION THREAD

**Discussion by thiippal** [explosion/spaCy#7872] [Help: Coding & Implementations] [answered]
Hey!

I'm trying to use the [`conllu_to_docs` function](https://github.com/explosion/spaCy/blob/master/spacy/training/converters/conllu_to_docs.py) to read CoNLL-U annotated files from the [Georgetown University Multilayer Corpus](https://corpling.uis.georgetown.edu/gum/index.html).

I start by reading the annotation file and pass the resulting string object to the `conllu_to_docs` function.

I then combine the list of Doc objects into a single Doc.

```python
# Open the plain text file for reading; assign under 'data'
data = open("data/GUM_whow_parachute.conllu", "r", encoding="utf-8")

# Read the file contents and assign under 'annotations'
annotations = data.read()

# Provide the string object under 'annotations' to the 'conllu_to_docs' function. 
# Set 'no_print' to True and cast the result into a Python list; store under 'docs'.
docs = list(conllu_to_docs(annotations, no_print=True))

# Combine Doc objects in the list 'docs' into a single Doc; assign under 'doc'
doc = Doc.from_docs(docs)
```

However, if I loop over some Tokens and print out the linguistic annotations, I get weird output, especially for the attribute `head`, which should contain indices for the heads in the syntactic relations.

```python
for token_ix in range(0, 8):
    
    token = doc[token_ix]
    
    print(token, token.tag_, token.pos_, token.morph, token.dep_, token.head)
```
```
How WRB SCONJ PronType=Int mark Cope
to TO PART  mark Cope
Cope VB VERB VerbForm=Inf ROOT Cope
With IN ADP  case Failure
a DT DET Definite=Ind|PronType=Art det Failure
Double JJ ADJ Degree=Pos amod Failure
Parachute NN NOUN Number=Sing compound Failure
Failure NN NOUN Number=Sing obl Cope
```

Any idea what causes this?

**Comment by adrianeboyd** [comment, marked as answer]
The value `Token.head` is a `Token` object from the doc, so when you print it you get the `repr` which is the `token.text`. You can use `token.head.i` to get the index of this token in the `Doc` (index in the 0-indexed `Doc`, not the sentence).

So to get back to CoNLL-U  you have to adjust it: `+1` for the first sentence in the doc and for later sentences adjusted by the index of the first word in the sentence `+ 1`

**Reply by thiippal** [reply]
Thanks @adrianeboyd for the quick reply!

Another question: is it possible to manually set the language of a *Doc* object? 

This information is naturally unavailable for Docs parsed from CoNLL-U, which prevents using e.g. the `noun_chunks` attribute.

**Reply by adrianeboyd** [reply]
You can't change the language of a doc because it's tied to the vocab used to create the doc, but you can inspect it under `nlp.vocab.lang`.

The converter uses a dummy vocab with no language set (which is not otherwise recommended), since normally you're saving `.spacy` `DocBin` files that also don't have a language associated with them.

Hmm, with the converter used in a script like this, as a workaround I think you can do a `DocBin` roundtrip where you use the intended vocab with `docbin.get_docs()` to regenerate the docs by passing the right vocab.

**Reply by thiippal** [reply]
Heh, I feel really stupid now – I thought the `head` attribute always returned the index of the head in spaCy, but this is obviously not true. Well, it's a Friday afternoon.

In a future version of spaCy, would it be possible to pass a *Vocab* object to the `conllu_to_docs` function as an alternative to the dummy *Vocab*, or would this break things?

**Reply by adrianeboyd** [reply]
Sure, I think in this case I would add the `model` parameter, which is already used for similar reasons in other converters. You can pass in `blank:lg` as the model name to get the same thing as `spacy.blank("en")`.
